lib/Corres_Cases.thy

```diff
@@ -0,0 +1,319 @@
+
+end
+
+
+section \<open>Examples and tests\<close>
+
+experiment
+begin
+
+(* abstract side *)
+lemma "corres_underlying srel nf nf' rrel (G x) G' (case x of None \<Rightarrow> a | Some y \<Rightarrow> b y) m"
+  (* produces strong (forall, implies) guard conditions in the final implications for both sides *)
+  apply corres_cases
+  oops
+
+schematic_goal
+  "\<And>x. corres_underlying srel nf nf' rrel (?G x) ?G' (case x of None \<Rightarrow> a | Some y \<Rightarrow> b y) m"
+  (* produces weak (just ?A \<and> ?B) guard conditions for concrete side, because ?G' does not
+     depend on "x", on which we do the case distinction *)
+  apply corres_cases
+  oops
+
+(* abstract side, with bind *)
+lemma "corres_underlying srel nf nf' rrel G G' ((case x of None \<Rightarrow> a | Some y \<Rightarrow> b y) >>= g) m"
+  apply corres_cases
+  oops
+
+(* abstract side, with bindE *)
+lemma "corres_underlying srel nf nf' rrel G G' ((case x of None \<Rightarrow> a | Some y \<Rightarrow> b y) >>=E g) m"
+  apply corres_cases
+  oops
+
+(* concrete side: *)
+lemma "corres_underlying srel nf nf' rrel G G' m (case x of None \<Rightarrow> a | Some y \<Rightarrow> b y)"
+  apply corres_cases
+  oops
+
+schematic_goal
+  "\<And>x. corres_underlying srel nf nf' rrel ?G (?G' x) m (case x of None \<Rightarrow> a | Some y \<Rightarrow> b y)"
+  apply corres_cases
+  oops
+
+lemma "corres_underlying srel nf nf' rrel G G' m ((case x of None \<Rightarrow> a | Some y \<Rightarrow> b y) >>= g)"
+  apply corres_cases
+  oops
+
+lemma "corres_underlying srel nf nf' rrel G G' m ((case x of None \<Rightarrow> a | Some y \<Rightarrow> b y) >>=E g)"
+  apply corres_cases
+  oops
+
+(* both sides: *)
+lemma "corres_underlying srel nf nf' rrel G G' (case x of None \<Rightarrow> a | Some y \<Rightarrow> b)
+                                               (case x of None \<Rightarrow> a' | Some y \<Rightarrow> b' y)"
+  (* two cases remain (both None, both Some); eliminated cases have guard \<top> in final implication *)
+  apply corres_cases_both
+  oops
+
+schematic_goal
+  "\<And>x y. corres_underlying srel nf nf' rrel (?G x) (?G' y) (case x of None \<Rightarrow> a | Some y \<Rightarrow> b)
+                                                           (case y of None \<Rightarrow> a' | Some y \<Rightarrow> b' y)"
+  (* 4 cases remain, because none are contradictory *)
+  apply corres_cases_both
+  oops
+
+(* some example relation between abstract and concrete values *)
+definition
+  "none_rel x y \<equiv> (x = None) = (y = None)"
+
+lemma
+  "none_rel x y \<Longrightarrow>
+   corres_underlying srel nf nf' rrel G G' (case x of None \<Rightarrow> a | Some y \<Rightarrow> b)
+                                            (case y of None \<Rightarrow> a' | Some y \<Rightarrow> b' y)"
+  (* two cases remain, none_rel is untouched in the cases that remain, but unfolded in the
+     ones that were eliminated *)
+  apply (corres_cases_both simp: none_rel_def)
+  oops
+
+end
```

Comment:
I'm never sure when we cross the threshold into needing tests to be placed in a separate file. This is short, and the extra definition (`none_rel`) shouldn't leak, so all good, but I'm curious as to your opinion on where that threshold lies.

Reply:
This is pretty close to that line. The section used to be much shorter, but then I introduced the >>E and then corres_cases_both and then the test with schematics and now it's 3 times as long..